packages/doenetml-worker-rust/lib-doenetml-core/src/new_core/prop_calculation.rs

```diff
@@ -0,0 +1,119 @@
+use crate::{
+    components::{
+        prelude::{ComponentState, DataQuery, FlatDastElementContent},
+        ComponentNode,
+    },
+    graph::directed_graph::Taggable,
+    state::{Freshness, PropPointer},
+    ComponentIdx,
+};
+
+use super::graph_based_core::Core;
+
+impl Core {
+    pub fn freshen_renderer_state(&mut self) -> Vec<ComponentIdx> {
+        let mut stale_renderers = Vec::new();
+
+        // recursively get a list of all rendered descendants of the components in stale_renderers
+        let mut stale_renderer_idx = 0;
+        while stale_renderer_idx < stale_renderers.len() {
+            let component_idx = stale_renderers[stale_renderer_idx];
+            stale_renderer_idx += 1;
+
+            stale_renderers.extend(
+                self.component_rendered_children(&self.components[component_idx])
+                    .into_iter()
+                    .filter_map(|child| match child {
+                        FlatDastElementContent::Text(_) => None,
+                        FlatDastElementContent::Element(child_idx) => Some(child_idx),
+                    }),
+            );
+        }
+
+        // deduplicate the list of stale renderers,
+        // sorting first since .dedup removes only consecutive duplicates.
+        stale_renderers.sort_unstable();
+        stale_renderers.dedup();
+
+        for component_idx in stale_renderers.iter() {
+            self.components[*component_idx].set_is_in_render_tree(true);
```

Comment:
It'd be nice to not mutate the components during render.

Reply:
Yes, I suspect we'll find another way to determine if it is in the render tree. I'll add a note. At this stage, I wasn't thinking about that problem.